drivers/fuel_gauge/bq27z746/emul_bq27z746.c

```diff
@@ -0,0 +1,352 @@
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "bq27z746.h"
+
+#define BQ27Z746_MAC_DATA_LEN     32
+#define BQ27Z746_MAC_OVERHEAD_LEN 4 /* 2 cmd bytes, 1 length byte, 1 checksum byte */
+#define BQ27Z746_MAC_COMPLETE_LEN (BQ27Z746_MAC_DATA_LEN + BQ27Z746_MAC_OVERHEAD_LEN)
+
+struct bq27z746_emul_data {
+	uint16_t mac_cmd;
+};
+
+/** Static configuration for the emulator */
+struct bq27z746_emul_cfg {
+	/** I2C address of emulator */
+	uint16_t addr;
+};
+
+static int emul_bq27z746_read_altmac(const struct emul *target, uint8_t *buf, size_t len)
+{
+	const uint8_t manufacturer_name[] = "Texas Instruments";
+	const uint8_t device_name[] = "BQ27Z746";
+	const uint8_t device_chemistry[] = "LION";
+	const struct bq27z746_emul_data *data = target->data;
+
+	if (len < BQ27Z746_MAC_COMPLETE_LEN) {
+		LOG_ERR("When reading the ALTMAC, one must read the full %u byte",
+			BQ27Z746_MAC_COMPLETE_LEN);
+		return -EIO;
+	}
+
+	memset(buf, 0, len);
+
+	/*
+	 * The data read from BQ27Z746_ALTMANUFACTURERACCESS is:
+	 * 0..1: The command (for verification)
+	 * 2..33: The data
+	 * 34: Checksum calculated as (uint8_t)(0xFF - (sum of all command and data bytes))
+	 * 35: Length including command, checksum and length (e.g. data length + 4)
+	 */
+
+	/* Put the command in the first two byte */
+	sys_put_le16(data->mac_cmd, buf);
+
+	/* Based on the command, put some data and the length into the buffer. */
+	/* In all of the operations, don't consider the zero-terminator. */
+	switch (data->mac_cmd) {
+	case BQ27Z746_MAC_CMD_MANUFACTURER_NAME:
+		memcpy(&buf[2], manufacturer_name, sizeof(manufacturer_name) - 1);
+		buf[35] = sizeof(manufacturer_name) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	case BQ27Z746_MAC_CMD_DEVICE_NAME:
+		memcpy(&buf[2], device_name, sizeof(device_name) - 1);
+		buf[35] = sizeof(device_name) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	case BQ27Z746_MAC_CMD_DEVICE_CHEM:
+		memcpy(&buf[2], device_chemistry, sizeof(device_chemistry) - 1);
+		buf[35] = sizeof(device_chemistry) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	default:
+		LOG_ERR("ALTMAC command 0x%x is not supported", data->mac_cmd);
+		return -EIO;
+	}
+
+	/* Calculate the checksum */
+	uint8_t sum = 0; /* Intentionally 8 bit wide and overflowing */
+
+	for (int i = 0; i < BQ27Z746_MAC_COMPLETE_LEN - 2; i++) {
+		sum += buf[i];
+	}
+	buf[34] = 0xFF - sum;
+
+	return 0;
+}
+
+static int emul_bq27z746_write(const struct emul *target, uint8_t *buf, size_t len)
+{
+	struct bq27z746_emul_data *data = target->data;
+	const uint8_t reg = buf[0];
+
+	switch (reg) {
+	case BQ27Z746_ALTMANUFACTURERACCESS:
+		data->mac_cmd = sys_get_le16(&buf[1]);
+		return 0;
+	default:
+		LOG_ERR("Writing is only supported to ALTMAC currently");
+		return -EIO;
+	}
+}
+
+static int emul_bq27z746_reg_read(const struct emul *target, int reg, int *val)
+{
+	switch (reg) {
+	case BQ27Z746_MANUFACTURERACCESS:
```

Comment:
Always thought these should return some unique value, like for instance:
case BQ27Z746_MANUFACTURERACCESS:
    *val = BQ27Z746_MANUFACTURERACCESS;
    break;

That way in the test, you could at least tell that the dispatcher was using the right symbol, however, it wouldn't solve the problem of then interpreting the data using the wrong union member - but it's a start.

Reply:
That's a good idea. @aaronemassey what do you think, shall I just return the register addresses?

Reply:
tldr: Using a unique magic stub value for that register would definitely be a better test.

In the long term what's really best is to add an [emulator backend API](https://docs.zephyrproject.org/latest/hardware/emulator/index.html) implementation that uses that register indirectly through some test scenaro.

Such as is being added to the https://github.com/zephyrproject-rtos/zephyr/pull/61435 battery cutoff work for SBS Gauge.

But that seems like a bit much for an initial PR of a driver at the moment. I agree with @perowle that using a different value is a better test.

Reply:
Well, I just noticed that the register defines are in the private header not currently accessible by the emulator. So returning the register values is easy, checking against them in the emulator code would mean to either put in magic numbers or put the register defines in a public header. I'm not convinced by either solution.

Could we solve this by a follow-up PR which introduces the emulator backend API?

Reply:
> Could we solve this by a follow-up PR which introduces the emulator backend API?

I think for testing purposes this is the best solution, especially in the long-term. Since emulator backends are relatively new, I'm also more than happy to help with that follow-up PR.